llvm/docs/LangRef.rst

```diff
@@ -19637,6 +19637,51 @@ are undefined.
     }
 
 
+'``llvm.experimental.vector.match.*``' Intrinsic
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+
+This is an overloaded intrinsic. Support for specific vector types is target
+dependent.
+
+::
+
+    declare <<n> x i1> @llvm.experimental.vector.match(<<n> x <ty>> %op1, <<n> x <ty>> %op2, <<n> x i1> %mask, i32 <segsize>)
+    declare <vscale x <n> x i1> @llvm.experimental.vector.match(<vscale x <n> x <ty>> %op1, <vscale x <n> x <ty>> %op2, <vscale x <n> x i1> %mask, i32 <segsize>)
+
+Overview:
+"""""""""
+
+Find elements of the first argument matching any elements of the second.
+
+Arguments:
+""""""""""
+
+The first argument is the search vector, the second argument is the vector of
+elements we are searching for (i.e. for which we consider a match successful),
+and the third argument is a mask that controls which elements of the first
+argument are active. The fourth argument is an immediate that sets the segment
+size for the search window.
+
+Semantics:
+""""""""""
+
+The '``llvm.experimental.vector.match``' intrinsic compares each element in the
+first argument against potentially several elements of the second, placing
+``1`` in the corresponding element of the output vector if any comparison is
+successful, and ``0`` otherwise. Inactive elements in the mask are set to ``0``
+in the output. The segment size controls the number of elements of the second
+argument that are compared against.
+
+For example, for vectors with 16 elements, if ``segsize = 16`` then each
+element of the first argument is compared against all 16 elements of the second
+argument; but if ``segsize = 4``, then each of the first four elements of the
+first argument is compared against the first four elements of the second
+argument, each of the second four elements of the first argument is compared
+against the second four elements of the second argument, and so forth.
+
```

Comment:
Is there a requirement that is an immediate value? If so, it might be worth documenting that, since it has an impact on scalable vectors where we don't know the total number of elements. For example, what's the expected behaviour if segsize > number of elements?

Alternatively, perhaps could take a value of -1 to indicate all elements should be searched?

Reply:
I made segsize an immediate to ease code generation, thinking that in the future we could lift the restriction if needed. Do you have a specific use case for a non-immediate segsize in mind?

For fixed-length vectors, I thought it would be easy to pass segsize = VL, but you are right, this can be problematic for scalable vectors if the hardware supports doing a match across the whole register. Accepting -1 to mean "all elements should be searched", as you suggested, sounds very reasonable. I'll work on that.

In general I'd rather not have segsize at all as it only complicates things, but adding it was the only way I found to accommodate the semantics of instructions such as SVE2's MATCH which operates on _segments_ of data (and not the whole register). Having said that, if you have a better strategy, I'm definitely keen to hear it and discuss it!